### PR TITLE
MONGOID-5128 Scoped associations

### PR DESCRIPTION
**Pull request by johnnyshields**
**Ready for review**

Fixes MONGOID-5128

ActiveRecord allows you to do this:

```ruby
class Author < ApplicationRecord
  has_many :books, -> { where(processed: true) }
end
```

I'm proposing equivalent syntax in Mongoid using `:scope` keyword arg.
If you prefer I could match the Rails syntax exactly, but I thought this was a bit clearer:

```ruby
class Author
  has_many :books, scope: -> { where(processed: true) }
end
```

In my implementation you can also use a Symbol to access a named scope:

```ruby
class Trainer
  has_many :pokemons, scope: :grass_type # equivalent to -> { grass_type }
end
```

I have some planned enhancements to this functionality but this PR is a good "minimum mergeable spec".

---

See:
- https://guides.rubyonrails.org/association_basics.html#scopes-for-belongs-to
- https://guides.rubyonrails.org/association_basics.html#scopes-for-has-one
- https://guides.rubyonrails.org/association_basics.html#scopes-for-has-many
- https://guides.rubyonrails.org/association_basics.html#scopes-for-has-and-belongs-to-many

### CODE REVIEW

**Comment by agolin95** [comment]
Tracked in https://jira.mongodb.org/browse/MONGOID-5128

**Comment by johnnyshields** [comment]
@agolin95 @p-mongo it would be great to get a code review on this when you have time. Please let me know if I'm headed in the right direction here.

**Comment by alcaeus** [comment]
@johnnyshields I've asked to take @comandeo to review this while @p-mongo is on well-deserved PTO. We'll get back to you once we've had a chance to take a look.

**Comment by johnnyshields** [comment]
OK, great. Yes, Evergreen being actually 🟢 green would be really helpful.

**Comment by johnnyshields** [comment]
Hmmm... so PR is progressing well, however, I'm running into a bit of a wall the more I think about my use case.

I'd like to do be able to define multiple scopes for the same association:

```ruby
class Game
  has_many :players
  has_many :active_players, scope: ->{ is_active: true }, class_name: 'Player', foreign_key: :game_id, inverse_of: nil
end
```

This adds a lot of boilerplate to achieve what I want. There's got to be a better way.

My first thought was to do it as an (extension)[https://docs.mongodb.com/mongoid/current/tutorials/mongoid-relations/#extensions], i.e.:

```ruby
class Game
  has_many :players do
    def active
      where(is_active: true).first
    end
  end
end
```

However, this won't work with eager loading. Perhaps it could be defined as a scope such as:

```ruby
  has_many :players do
    scope :active_players, ->{ is_active: true }
  end
```
  
And do eager loading by callling `Event.games.includes(:active_players)` 
  
Alternatively, we could do:

```ruby
class Game
  has_many :players
  has_many :active_players, extends: :players, scope: ->{ is_active: true }
  
  # or
  association_scope :active_players, :players, ->{ is_active: true }
end
```

`association_scope` may be the best. It may actually make sense to introduce a new relationship class, perhaps `Mongoid::Association::Scoped`? Ideally these association_scopes are "read-only", or any assignment would be delegated to the underlying association.

**Comment by johnnyshields** [comment]
I've implemented `association_scope` (Option C) from above.

**Comment by johnnyshields** [comment]
@p-mongo would like to get your comments on this.

**Comment by p-mongo** [comment]
I support implementing the AR-compatible syntax for defining the scope criteria.

Mongoid (and I believe AR too) already provides scoping under associations:
```
class Foo
  include Mongoid::Document
  
  has_many :bars
end

class Bar
  include Mongoid::Document

  scope :open, -> { where(open: true) }
end

irb(main):012:0> Foo.create!
=> #<Foo _id: 6109526fa15d5d20c601e23d, >
irb(main):013:0> Foo.first.bars
=> []
irb(main):014:0> Foo.first.bars.open
=> 
#<Mongoid::Criteria
  selector: {"_id"=>BSON::ObjectId('6109526fa15d5d20c601e23d'), "open"=>true}
  options:  {}
  class:    Bar
  embedded: false>

```

**Comment by johnnyshields** [comment]
> Mongoid (and I believe AR too) already provides scoping under associations

Yes, BUT there is one important point. You can't *eager load* scoped associations.

In other words:

```
# you can't do this
Foo.all.includes(:"bars.open")

# my PR allows this
class Foo
  ...
  association_scope :open_bars, :bars, -> { where(open: true) }
end

Foo.all.includes(:open_bars)
```

**Comment by p-mongo** [comment]
I believe that pattern is already possible to implement in Mongoid simply by defining additional associations for each needed scope and setting `class_name` appropriately.

**Comment by johnnyshields** [comment]
Hmmm.... I'm not following. Can you provide a code (or pseudo-code) example?

**Comment by johnnyshields** [comment]
@p-mongo this is now ready for review. I've dropped the `association_scope` function, can be considered in the future after more investigation.

File: spec/mongoid/association/referenced/belongs_to/buildable_spec.rb [hunk 1]
```diff
@@ -14,31 +14,70 @@
...
+          Person.create(_id: object, username: 'Bob')
         end
 
         let(:object) do
           BSON::ObjectId.new
         end
 
         before do
-          expect(Person).to receive(:where).with(association.primary_key => object).and_call_original
+          expect_any_instance_of(Mongoid::Criteria).to receive(:where).with(association.primary_key => object).and_call_original
```
Comment (p-mongo) [inline spec/mongoid/association/referenced/belongs_to/buildable_spec.rb:38]: What is the purpose of this and similar changes throughout the PR please?

**Comment by p-mongo** [comment]
Overall this looks to be very thorough, thank you for this PR.

AR defines its association thusly:

```
activerecord/lib/active_record/associations.rb:      def has_and_belongs_to_many(name, scope = nil, options = {}, &extension)
```
The scope argument was added in 4.0. Previously the definition looked like this:
```
activerecord/lib/active_record/associations.rb:      def has_and_belongs_to_many(name, options = {}, &extension)
```

I imagine such a change would be a breaking one, hence I see the value in using a keyword argument for scope. At the same time it should be relatively straightforward to support both invocations by examining the type of `scope` and if it's a Hash, treating it as options. It is my impression that this compatibility mechanism would continue to work even with the double splat syntax and Ruby 3 which is what Rails currently has in its codebase:
```
activerecord/lib/active_record/associations.rb:        def has_and_belongs_to_many(name, scope = nil, **options, &extension)
```

File: spec/mongoid/association/referenced/belongs_to/buildable_spec.rb (hunk 1 above)
Comment (johnnyshields) [reply spec/mongoid/association/referenced/belongs_to/buildable_spec.rb:38]: It's related to this change in buildable:

```ruby
          def query_criteria(object, base)
            crit = klass.criteria                        # <== this line
            crit = crit.apply_scope(scope)
            crit = crit.where(foreign_key => object)
            with_polymorphic_criterion(crit, base)
          end
```

We're now explicitly creating the criteria first line, so the call doesn't happen on the Person class, it happens on a Criteria object.

**Comment by johnnyshields** [comment]
@p-mongo I am intentionally not matching ActiveRecord's "second arg" syntax. `scope` will be used relatively rarely, there is no reason to give it such premium treatment.

**Comment by johnnyshields** [comment]
@comandeo ready for your review.

**Comment by p-mongo** [comment]
@comandeo What is your opinion on the question of whether Mongoid should match AR behavior and accept the scope as a positional argument? I see three options at this point:

1.

Positional argument only, matches AR behavior.


2.

Keyword argument only. This is the current implementation in this PR. Not compatible with AR in either direction.

3.

Both keyword argument and positional argument are allowed. Compatible with AR in that AR-targeting code will work with Mongoid, but permits also specifying all association options as keyword arguments which I can see how this can be more consistent.

**Comment by comandeo** [comment]
I think in this case it is better to be consistent and treat all arguments similarly. I agree that we should be as close as possible to AR in general; however, supporting both ways here won't bring us benefits, actually.

**Comment by johnnyshields** [comment]
@comandeo the AR syntax for this option in is quite ugly. It really doesn't make sense for `scope` to be a second argument, especially because it comes with significant caveats. I'd like to keep it as I have in this PR.

**Comment by johnnyshields** [comment]
Well hmmm... on second thought its ok. I guess we can keep consistent with AR since all the other keys match exactly. I'll change it.

**Comment by johnnyshields** [comment]
I take back my previous comment. Having it as a positional arg makes the code messier. Let's keep it as a keyword arg.

**Comment by comandeo** [comment]
I think we should merge this as soon as CI is green. @johnnyshields I gave the Evergreen a kick, and all configurations (except app tests) fail with the following:

```
   1) Mongoid::Criteria::Scopable#apply_scope when the scope is a Criteria when standard scope merges the criteria
       Failure/Error: raise Errors::InvalidIncludes.new(_parent_class, [ relation_object ]) unless association
       
       Mongoid::Errors::InvalidIncludes:
       
         message:
           Invalid includes directive: Band.includes(:drugs)
         summary:
           Eager loading in Mongoid only supports providing arguments to Band.includes that are the names of associations on the Band.
         resolution:
           Ensure that each parameter passed to Band.includes is a valid name of an association on the Band model. These are: "records", "notes", "labels", "label", "same_name".
       # ./lib/mongoid/criteria/includable.rb:81:in `block in extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:71:in `each'
       # ./lib/mongoid/criteria/includable.rb:71:in `extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:29:in `includes'
       # ./spec/mongoid/criteria/scopable_spec.rb:135:in `block (5 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:139:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
 
 
    2) Mongoid::Criteria::Scopable#apply_scope when the scope is a Criteria when unscoped unscoped has no effect
       Failure/Error: raise Errors::InvalidIncludes.new(_parent_class, [ relation_object ]) unless association
       
       Mongoid::Errors::InvalidIncludes:
       
         message:
           Invalid includes directive: Band.includes(:drugs)
         summary:
           Eager loading in Mongoid only supports providing arguments to Band.includes that are the names of associations on the Band.
         resolution:
           Ensure that each parameter passed to Band.includes is a valid name of an association on the Band model. These are: "records", "notes", "labels", "label", "same_name".
       # ./lib/mongoid/criteria/includable.rb:81:in `block in extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:71:in `each'
       # ./lib/mongoid/criteria/includable.rb:71:in `extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:29:in `includes'
       # ./spec/mongoid/criteria/scopable_spec.rb:147:in `block (5 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:151:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
 
 
    3) Mongoid::Criteria::Scopable#apply_scope when the scope is a Proc when standard scope adds the scope
       Failure/Error: raise Errors::InvalidIncludes.new(_parent_class, [ relation_object ]) unless association
       
       Mongoid::Errors::InvalidIncludes:
       
         message:
           Invalid includes directive: Band.includes(:drugs)
         summary:
           Eager loading in Mongoid only supports providing arguments to Band.includes that are the names of associations on the Band.
         resolution:
           Ensure that each parameter passed to Band.includes is a valid name of an association on the Band model. These are: "records", "notes", "labels", "label", "same_name".
       # ./lib/mongoid/criteria/includable.rb:81:in `block in extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:71:in `each'
       # ./lib/mongoid/criteria/includable.rb:71:in `extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:29:in `includes'
       # ./spec/mongoid/criteria/scopable_spec.rb:162:in `block (6 levels) in <top (required)>'
       # ./lib/mongoid/criteria/scopable.rb:35:in `instance_exec'
       # ./lib/mongoid/criteria/scopable.rb:35:in `apply_scope'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:166:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
 
 
    4) Mongoid::Criteria::Scopable#apply_scope when the scope is a Proc when unscoped removes existing scopes then adds the new scope
       Failure/Error: raise Errors::InvalidIncludes.new(_parent_class, [ relation_object ]) unless association
       
       Mongoid::Errors::InvalidIncludes:
       
         message:
           Invalid includes directive: Band.includes(:drugs)
         summary:
           Eager loading in Mongoid only supports providing arguments to Band.includes that are the names of associations on the Band.
         resolution:
           Ensure that each parameter passed to Band.includes is a valid name of an association on the Band model. These are: "records", "notes", "labels", "label", "same_name".
       # ./lib/mongoid/criteria/includable.rb:81:in `block in extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:71:in `each'
       # ./lib/mongoid/criteria/includable.rb:71:in `extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:29:in `includes'
       # ./spec/mongoid/criteria/scopable_spec.rb:174:in `block (6 levels) in <top (required)>'
       # ./lib/mongoid/criteria/scopable.rb:35:in `instance_exec'
       # ./lib/mongoid/criteria/scopable.rb:35:in `apply_scope'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:178:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
 
 
    5) Mongoid::Criteria::Scopable#apply_scope when the scope is a Symbol when standard scope adds the scope
       Failure/Error: super
       
       NoMethodError:
         undefined method `highly_rated' for #<Mongoid::Criteria
           selector: {"name"=>"Black Sabbath"}
           options:  {:skip=>20}
           class:    Band
           embedded: false>
       # ./lib/mongoid/criteria.rb:529:in `method_missing'
       # ./lib/mongoid/criteria/scopable.rb:37:in `apply_scope'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:193:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
 
 
    6) Mongoid::Criteria::Scopable#apply_scope when model has default_scope when standard scope merges the scope
       Failure/Error: raise Errors::InvalidIncludes.new(_parent_class, [ relation_object ]) unless association
       
       Mongoid::Errors::InvalidIncludes:
       
         message:
           Invalid includes directive: Band.includes(:drugs)
         summary:
           Eager loading in Mongoid only supports providing arguments to Band.includes that are the names of associations on the Band.
         resolution:
           Ensure that each parameter passed to Band.includes is a valid name of an association on the Band model. These are: "records", "notes", "labels", "label", "same_name".
       # ./lib/mongoid/criteria/includable.rb:81:in `block in extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:71:in `each'
       # ./lib/mongoid/criteria/includable.rb:71:in `extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:29:in `includes'
       # ./spec/mongoid/criteria/scopable_spec.rb:224:in `block (6 levels) in <top (required)>'
       # ./lib/mongoid/criteria/scopable.rb:35:in `instance_exec'
       # ./lib/mongoid/criteria/scopable.rb:35:in `apply_scope'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:228:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
 
 
    7) Mongoid::Criteria::Scopable#apply_scope when model has default_scope when unscoped removes existing scopes then adds the new scope
       Failure/Error: raise Errors::InvalidIncludes.new(_parent_class, [ relation_object ]) unless association
       
       Mongoid::Errors::InvalidIncludes:
       
         message:
           Invalid includes directive: Band.includes(:drugs)
         summary:
           Eager loading in Mongoid only supports providing arguments to Band.includes that are the names of associations on the Band.
         resolution:
           Ensure that each parameter passed to Band.includes is a valid name of an association on the Band model. These are: "records", "notes", "labels", "label", "same_name".
       # ./lib/mongoid/criteria/includable.rb:81:in `block in extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:71:in `each'
       # ./lib/mongoid/criteria/includable.rb:71:in `extract_includes_list'
       # ./lib/mongoid/criteria/includable.rb:29:in `includes'
       # ./spec/mongoid/criteria/scopable_spec.rb:236:in `block (6 levels) in <top (required)>'
       # ./lib/mongoid/criteria/scopable.rb:35:in `instance_exec'
       # ./lib/mongoid/criteria/scopable.rb:35:in `apply_scope'
       # ./spec/mongoid/criteria/scopable_spec.rb:128:in `block (3 levels) in <top (required)>'
       # ./spec/mongoid/criteria/scopable_spec.rb:240:in `block (5 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
       # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
```
It looks like something is missing there. Could you please take a look? Thank you!

**Comment by johnnyshields** [comment]
@comandeo fixed!

**Review by comandeo — APPROVED** [review]

**Comment by p-mongo** [comment]
I removed the link to named scopes because it was broken and I couldn't find anywhere in our docs where we actually document them. @johnnyshields where was that supposed to go?

**Comment by johnnyshields** [comment]
Named scopes was supposed to link to here: https://docs.mongodb.com/mongoid/current/tutorials/mongoid-queries/#named-scopes

**Pull request merged**